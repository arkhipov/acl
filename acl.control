# ACL extension
comment = 'data type ACL'
default_version = '1.0.3'
module_pathname = '$libdir/acl'
relocatable = true
