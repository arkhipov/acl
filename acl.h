/* -------------------------------------------------------------------------
 *
 * acl.h
 *
 * Copyright (c) 2015-2016 Vladislav Arkhipov <vlad@arkhipov.ru>
 *
 * -------------------------------------------------------------------------
 */
#ifndef __ACL_H__
#define __ACL_H__

#include "lib/stringinfo.h"
#include "utils/array.h"

/* ACE types */
#define ACE_ACCESS_ALLOWED				0x00000001
#define ACE_ACCESS_DENIED				0x00000002

/* ACE flags
 * 0-15 application-specific
 * 16-31 reserved */
#define ACE_INHERIT_ONLY				0x80000000
#define ACE_OBJECT_INHERIT				0x40000000
#define ACE_CONTAINER_INHERIT			0x20000000
#define ACE_NO_PROPAGATE_INHERIT		0x10000000
#define ACE_INHERITED					0x08000000
#define ACE_INVALID						0x04000000

#define ACE_FLAGS_APPLICATION_SPECIFIC	0x0000FFFF

/* ACE access rights
 * 0-15 application-specific
 * 16-31 reserved */
#define ACE_READ						0x80000000
#define ACE_WRITE						0x40000000
#define ACE_DELETE						0x20000000
#define ACE_READ_ACL					0x10000000
#define ACE_WRITE_ACL					0x08000000

typedef struct AclEntryBase {
	uint32			type;
	uint32			flags;
	uint32			mask;
} AclEntryBase;

void parse_acl_entry(const char *s, AclEntryBase *acl_entry_base,
					 void *opaque,
					 const char *parse_who(const char *s, void *opaque));

void format_acl_entry(StringInfo out, intptr_t opaque,
					  AclEntryBase *acl_entry_base,
					  void (*format_who)(StringInfo out, intptr_t opaque));

uint32 check_access(const ArrayType *acl, int16 typlen, char typalign,
					AclEntryBase * (*extract_acl_entry_base)(void *acl_entry),
					uint32 mask, intptr_t who,
					bool (*who_matches)(void *acl_entry, intptr_t who),
					bool implicit_allow);

text *check_access_text_mask(const ArrayType *acl, int16 typlen,
							 char typalign,
							 AclEntryBase * (*extract_acl_entry_base)(void *acl_entry),
							 text *text_mask, intptr_t who,
							 bool (*who_matches)(void *acl_entry, intptr_t who),
							 bool implicit_allow);

ArrayType *merge_acls(const ArrayType *parent_acl, const ArrayType *acl,
					  int16 typlem, char typalign,
					  AclEntryBase * (*extract_acl_entry_base)(void *acl_entry),
					  bool container, bool deny_first);

#endif
