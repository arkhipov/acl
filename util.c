/* -------------------------------------------------------------------------
 *
 * util.c
 *
 * Copyright (c) 2015-2023 Vladislav Arkhipov <vlad@arkhipov.ru>
 *
 * -------------------------------------------------------------------------
 */
#include "util.h"

static void check_who_array(ArrayType *who_array);

bool
check_access_extract_args(FunctionCallInfo fcinfo, ArrayType **acl,
						  uint32 *mask, ArrayType **who, bool *implicit_allow,
						  bool extract_who, bool has_who_argument)
{
	if (PG_ARGISNULL(0))
		*acl = NULL;
	else
		*acl = PG_GETARG_ARRAYTYPE_P(0);

	if (PG_ARGISNULL(1))
		return false;

	*mask = PG_GETARG_UINT32(1);

	if (has_who_argument && PG_ARGISNULL(2))
		return false;

	if (extract_who)
	{
		*who = PG_GETARG_ARRAYTYPE_P(2);
		check_who_array(*who);
	}

	if (PG_ARGISNULL(has_who_argument ? 3 : 2))
		ereport(ERROR,
				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
				 errmsg("allow_implicit argument must be not null")));

	*implicit_allow = PG_GETARG_BOOL(3);

	return true;
}

bool
check_access_text_mask_extract_args(FunctionCallInfo fcinfo, ArrayType **acl,
									text **mask, ArrayType **who,
									bool *implicit_allow, bool extract_who,
									bool has_who_argument)
{
	if (PG_ARGISNULL(0))
		*acl = NULL;
	else
		*acl = PG_GETARG_ARRAYTYPE_P(0);

	if (PG_ARGISNULL(1))
		return false;

	*mask = PG_GETARG_TEXT_P(1);

	if (has_who_argument && PG_ARGISNULL(2))
		return false;

	if (extract_who)
	{
		*who = PG_GETARG_ARRAYTYPE_P(2);
		check_who_array(*who);
	}

	if (PG_ARGISNULL(has_who_argument ? 3 : 2))
		ereport(ERROR,
				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
				 errmsg("allow_implicit argument must be not null")));

	*implicit_allow = PG_GETARG_BOOL(3);

	return true;
}

void
merge_acls_extract_args(FunctionCallInfo fcinfo, ArrayType **parent,
						ArrayType **child, bool *container, bool *deny_first)
{
	if (PG_ARGISNULL(0))
		*parent = NULL;
	else
		*parent = PG_GETARG_ARRAYTYPE_P(0);

	if (PG_ARGISNULL(1))
		ereport(ERROR,
				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
				 errmsg("ACL must be not null")));

	*child = PG_GETARG_ARRAYTYPE_P(1);

	if (PG_ARGISNULL(2))
		ereport(ERROR,
				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
				 errmsg("container argument must be not null")));

	*container = PG_GETARG_BOOL(2);

	if (PG_ARGISNULL(3))
		ereport(ERROR,
				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
				 errmsg("deny_first argument must be not null")));

	*deny_first = PG_GETARG_BOOL(3);
}

static void
check_who_array(ArrayType *who_array)
{
	if (ARR_HASNULL(who_array))
		ereport(ERROR,
				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
				 errmsg("Who array must not contain null values")));

	if (ARR_NDIM(who_array) > 1)
		ereport(ERROR,
				(errcode(ERRCODE_INTERNAL_ERROR),
				 errmsg("wrong number of dimensions of who array"),
				 errdetail("Who array must be one dimensional.")));

	if (ARR_NDIM(who_array) > 0 && ARR_LBOUND(who_array)[0] != 1)
		ereport(ERROR,
				(errcode(ERRCODE_INTERNAL_ERROR),
				 errmsg("wrong range of who array"),
				 errdetail("Lower bound of who array must be one.")));
}
