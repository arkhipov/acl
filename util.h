/* -------------------------------------------------------------------------
 *
 * util.h
 *
 * Copyright (c) 2015-2023 Vladislav Arkhipov <vlad@arkhipov.ru>
 *
 * -------------------------------------------------------------------------
 */
#ifndef __UTIL_H__
#define __UTIL_H__

#include "postgres.h"
#include "fmgr.h"

#include "utils/array.h"

bool check_access_extract_args(FunctionCallInfo fcinfo, ArrayType **acl,
							   uint32 *mask, ArrayType **who,
							   bool *implicit_allow, bool extract_who,
							   bool has_who_argument);
bool check_access_text_mask_extract_args(FunctionCallInfo fcinfo,
										 ArrayType **acl, text **mask,
										 ArrayType **who, bool *implicit_allow,
										 bool extract_who,
										 bool has_who_argument);

void merge_acls_extract_args(FunctionCallInfo fcinfo, ArrayType **parent,
							 ArrayType **child, bool *container,
							 bool *deny_first);

#endif
