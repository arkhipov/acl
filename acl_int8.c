/* -------------------------------------------------------------------------
 *
 * acl_int8.c
 *
 * Copyright (c) 2015-2023 Vladislav Arkhipov <vlad@arkhipov.ru>
 *
 * -------------------------------------------------------------------------
 */
#include "postgres.h"
#include "fmgr.h"

#include "utils/builtins.h"
#include "utils/int8.h"

#include "acl.h"
#include "util.h"

PGDLLEXPORT Datum ace_int8_in(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum ace_int8_out(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_int8_check_access_text(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_int8_check_access_int4(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_int8_merge(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(ace_int8_in);
PG_FUNCTION_INFO_V1(ace_int8_out);
PG_FUNCTION_INFO_V1(acl_int8_check_access_text);
PG_FUNCTION_INFO_V1(acl_int8_check_access_int4);
PG_FUNCTION_INFO_V1(acl_int8_merge);

typedef struct AclEntryInt8
{
	AclEntryBase 	base;
	char			who[8];
} AclEntryInt8;

#define ACL_TYPE_ALIGNMENT				'i'
#define ACL_TYPE_LENGTH					sizeof(AclEntryInt8)

#define DatumGetInt8AclEntryP(x)		((AclEntryInt8 *) DatumGetPointer(x))
#define PG_GETARG_BIGINT_ACL_ENTRY_P(x)	DatumGetInt8AclEntryP(PG_GETARG_DATUM(x))
#define PG_RETURN_BIGINT_ACL_ENTRY_P(x)	PG_RETURN_POINTER(x)

static const char *parse_who(const char *s, void *opaque);
static void format_who(StringInfo out, intptr_t acl_entry);

static AclEntryBase *extract_acl_entry_base(void *entry);
static bool who_matches(void *entry, intptr_t who);

Datum
ace_int8_in(PG_FUNCTION_ARGS)
{
	const char	   *s = PG_GETARG_CSTRING(0);
	AclEntryInt8   *entry;

	entry = palloc0(sizeof(AclEntryInt8));

	parse_acl_entry(s, &entry->base, entry->who, parse_who);

	PG_RETURN_BIGINT_ACL_ENTRY_P(entry);
}

Datum
ace_int8_out(PG_FUNCTION_ARGS)
{
	AclEntryInt8   *entry = PG_GETARG_BIGINT_ACL_ENTRY_P(0);
	StringInfo		out;

	out = makeStringInfo();

	format_acl_entry(out, (intptr_t) entry->who, &entry->base, format_who);

	PG_RETURN_CSTRING(out->data);
}

Datum
acl_int8_check_access_int4(PG_FUNCTION_ARGS)
{
	ArrayType	   *acl;
	uint32			mask;
	ArrayType	   *who;
	bool			implicit_allow;
	uint32			result;

	if (!check_access_extract_args(fcinfo, &acl, &mask, &who, &implicit_allow,
								   true, true))
		PG_RETURN_NULL();

	result = check_access(acl, ACL_TYPE_LENGTH, ACL_TYPE_ALIGNMENT,
						  extract_acl_entry_base, mask,
						  (intptr_t) who, who_matches,
						  implicit_allow);

	PG_RETURN_UINT32(result);
}

Datum
acl_int8_check_access_text(PG_FUNCTION_ARGS)
{
	ArrayType	   *acl;
	text		   *mask;
	ArrayType	   *who;
	bool			implicit_allow;
	text		   *result;

	if (!check_access_text_mask_extract_args(fcinfo, &acl, &mask, &who,
											 &implicit_allow, true, true))
		PG_RETURN_NULL();

	result = check_access_text_mask(acl, ACL_TYPE_LENGTH,
									ACL_TYPE_ALIGNMENT,
									extract_acl_entry_base, mask,
									(intptr_t) who, who_matches,
									implicit_allow);

	PG_RETURN_TEXT_P(result);
}

Datum
acl_int8_merge(PG_FUNCTION_ARGS)
{
	ArrayType	   *parent;
	ArrayType	   *child;
	bool			container;
	bool			deny_first;

	merge_acls_extract_args(fcinfo, &parent, &child, &container, &deny_first);

	PG_RETURN_ARRAYTYPE_P(merge_acls(parent, child,
									 ACL_TYPE_LENGTH, ACL_TYPE_ALIGNMENT,
									 extract_acl_entry_base,
									 container, deny_first));
}

static const char *
parse_who(const char *s, void *opaque)
{
	char			str[21];
	int				len = 0;
	int64			who;

	for (; *s != '\0' && (*s == '-' || isdigit((unsigned char) *s)); ++s)
	{
		if (len >= 20)
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_TEXT_REPRESENTATION),
					 errmsg("int8 too long")));

		str[len++] = *s;
	}

	str[len] = '\0';

	who = DatumGetInt64(DirectFunctionCall1(int8in, CStringGetDatum(str)));
	memcpy(opaque, &who, 8);

	return s;
}

static void
format_who(StringInfo out, intptr_t opaque)
{
	int64 who;

	memcpy(&who, (void *) opaque, 8);

	appendStringInfoString(out, DatumGetCString(DirectFunctionCall1(
										int8out, Int64GetDatum(who))));
}

static AclEntryBase *
extract_acl_entry_base(void *entry)
{
	return &((AclEntryInt8 *) entry)->base;
}

static bool
who_matches(void *entry, intptr_t who)
{
	int64			entry_who;
	bool			result = false;
	int				i, num;
	int64		   *ptr;

	memcpy(&entry_who, ((AclEntryInt8 *) entry)->who, 8);

	num = ArrayGetNItems(ARR_NDIM((ArrayType *) who),
						 ARR_DIMS((ArrayType *) who));
	ptr = (int64 *) ARR_DATA_PTR((ArrayType *) who);

	for (i = 0; i < num; ++i)
	{
		int64			who_value = *ptr++;

		if (entry_who == who_value)
		{
			result = true;
			break;
		}
	}

	return result;
}
