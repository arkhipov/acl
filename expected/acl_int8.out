-- ACL format checks
select ''::ace_int8;
ERROR:  missing ACE type
LINE 1: select ''::ace_int8;
               ^
select 'a'::ace_int8;
ERROR:  missing "/" sign
LINE 1: select 'a'::ace_int8;
               ^
select 'q'::ace_int8;
ERROR:  invalid ACE type: must be one of "ad"
LINE 1: select 'q'::ace_int8;
               ^
select 'a*'::ace_int8;
ERROR:  missing "/" sign
LINE 1: select 'a*'::ace_int8;
               ^
select 'a/'::ace_int8;
ERROR:  missing ACE flags
LINE 1: select 'a/'::ace_int8;
               ^
select 'a/hq'::ace_int8;
ERROR:  invalid ACE flag: must be one of "hpcoi0123456789ABCDEFGHIJKLMNOP"
LINE 1: select 'a/hq'::ace_int8;
               ^
select 'a/h'::ace_int8;
ERROR:  missing "/" sign
LINE 1: select 'a/h'::ace_int8;
               ^
select 'a/h/'::ace_int8;
ERROR:  missing ACE who
LINE 1: select 'a/h/'::ace_int8;
               ^
select 'a/h/='::ace_int8;
ERROR:  invalid input syntax for integer: ""
LINE 1: select 'a/h/='::ace_int8;
               ^
select 'a/h/1='::ace_int8;
ERROR:  missing ACE mask
LINE 1: select 'a/h/1='::ace_int8;
               ^
select 'a/h/1=d,'::ace_int8;
ERROR:  invalid ACE mask: must be one of "scdwr0123456789ABCDEFGHIJKLMNOPQ"
LINE 1: select 'a/h/1=d,'::ace_int8;
               ^
select 'a/h/1=dw'::ace_int8;
 ace_int8 
----------
 a/h/1=dw
(1 row)

select 'a/ihpc/1=wdddw'::ace_int8;
  ace_int8   
-------------
 a/hpci/1=dw
(1 row)

select 'd/ihpc/1=wdddw'::ace_int8;
  ace_int8   
-------------
 d/hpci/1=dw
(1 row)

select 'd/ihpc/-9223372036854775808=wdddw'::ace_int8;
            ace_int8            
--------------------------------
 d/hpci/-9223372036854775808=dw
(1 row)

select 'd/ihpc/9223372036854775807=wdddw'::ace_int8;
           ace_int8            
-------------------------------
 d/hpci/9223372036854775807=dw
(1 row)

select 'a/ihpc/blah"=dw0'::ace_int8;
ERROR:  invalid input syntax for integer: ""
LINE 1: select 'a/ihpc/blah"=dw0'::ace_int8;
               ^
select 'd/ihpc/9223372036854775808=wdddw'::ace_int8;
ERROR:  value "9223372036854775808" is out of range for type bigint
LINE 1: select 'd/ihpc/9223372036854775808=wdddw'::ace_int8;
               ^
select 'd/ihpc/922337203685477580800=wdddw'::ace_int8;
ERROR:  int8 too long
LINE 1: select 'd/ihpc/922337203685477580800=wdddw'::ace_int8;
               ^
-- check access
select coalesce(acl_check_access('{}'::ace_int8[], 'sd0', '{3, 2}'::int8[], false), 'NULL');
 coalesce 
----------
 
(1 row)

select coalesce(acl_check_access('{}'::ace_int8[], 'sd0', '{3, 2}'::int8[], true), 'NULL');
 coalesce 
----------
 0sd
(1 row)

select coalesce(acl_check_access(null::ace_int8[], 'sd0', '{3, 2}'::int8[], false), 'NULL');
 coalesce 
----------
 
(1 row)

select coalesce(acl_check_access(null::ace_int8[], 'sd0', '{3, 2}'::int8[], true), 'NULL');
 coalesce 
----------
 0sd
(1 row)

select coalesce(acl_check_access(null::ace_int8[], 'sd0', '{}'::int8[], false), 'NULL');
 coalesce 
----------
 
(1 row)

select coalesce(acl_check_access(null::ace_int8[], 'sd0', '{}'::int8[], true), 'NULL');
 coalesce 
----------
 0sd
(1 row)

select coalesce(acl_check_access(null::ace_int8[], 'sd0', null::int8[], false), 'NULL');
 coalesce 
----------
 NULL
(1 row)

select coalesce(acl_check_access(null::ace_int8[], 'sd0', null::int8[], true), 'NULL');
 coalesce 
----------
 NULL
(1 row)

select acl_check_access('{}'::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{3, 2}'::int8[], false)::bit(32);
         acl_check_access         
----------------------------------
 00000000000000000000000000000000
(1 row)

select acl_check_access('{}'::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{3, 2}'::int8[], true)::bit(32);
         acl_check_access         
----------------------------------
 00101000000000000000000000000001
(1 row)

select acl_check_access(null::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{3, 2}'::int8[], false)::bit(32);
         acl_check_access         
----------------------------------
 00000000000000000000000000000000
(1 row)

select acl_check_access(null::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{3, 2}'::int8[], true)::bit(32);
         acl_check_access         
----------------------------------
 00101000000000000000000000000001
(1 row)

select acl_check_access(null::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{}'::int8[], false)::bit(32);
         acl_check_access         
----------------------------------
 00000000000000000000000000000000
(1 row)

select acl_check_access(null::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{}'::int8[], true)::bit(32);
         acl_check_access         
----------------------------------
 00101000000000000000000000000001
(1 row)

select acl_check_access(null::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), null::int8[], false)::bit(32);
 acl_check_access 
------------------
 
(1 row)

select acl_check_access(null::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), null::int8[], true)::bit(32);
 acl_check_access 
------------------
 
(1 row)

select coalesce(acl_check_access('{d//1=w,d//2=s,a//2=sdw,a//3=0}'::ace_int8[], 'sdc0', '{3, 2}'::int8[], false), 'NULL');
 coalesce 
----------
 0d
(1 row)

select coalesce(acl_check_access('{d//1=w,d//2=s,a//2=sdw,a//3=0}'::ace_int8[], 'sdc0', '{3, 2}'::int8[], true), 'NULL');
 coalesce 
----------
 0cd
(1 row)

select acl_check_access('{d//1=w,d//2=s,a//2=sdw,a//3=0}'::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{3, 2}'::int8[], false)::bit(32);
         acl_check_access         
----------------------------------
 00100000000000000000000000000001
(1 row)

select acl_check_access('{d//1=w,d//2=s,a//2=sdw,a//3=0}'::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{3, 2}'::int8[], true)::bit(32);
         acl_check_access         
----------------------------------
 00100000000000000000000000000001
(1 row)

select coalesce(acl_check_access(null::ace_int8[], 'sd0', '{}'::int8[], null), 'NULL');
ERROR:  allow_implicit argument must be not null
select acl_check_access(null::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{}'::int8[], null)::bit(32);
ERROR:  allow_implicit argument must be not null
-- inherit only
select acl_check_access('{d//1=w,d/i/2=s,a//2=sdw,a//3=0}'::ace_int8[], 'sd0', '{3, 2}'::int8[], false);
 acl_check_access 
------------------
 0sd
(1 row)

select acl_check_access('{d//1=w,d/i/2=s,a//2=sdw,a//3=0}'::ace_int8[], (1 << 0) | (1 << 27) | (1 << 29), '{3, 2}'::int8[], false)::bit(32);
         acl_check_access         
----------------------------------
 00101000000000000000000000000001
(1 row)

-- merge
select acl_merge(null::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

select acl_merge(null::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, true);
           acl_merge            
--------------------------------
 {d//0=1,d//0=4,a//0=0,a//0=23}
(1 row)

select acl_merge(null::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

select acl_merge(null::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, true);
           acl_merge            
--------------------------------
 {d//0=1,d//0=4,a//0=0,a//0=23}
(1 row)

-- inheritance
-- container
-- no flags -> not inherited
select acl_merge('{a//1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

-- inherit only -> not inherited
select acl_merge('{a/i/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

-- object inherit -> inherit only + object inherit
select acl_merge('{a/o/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
                acl_merge                 
------------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/hoi/1=d}
(1 row)

-- object inherit + inherit only -> inherit only + object inherit
select acl_merge('{a/oi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
                acl_merge                 
------------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/hoi/1=d}
(1 row)

-- object inherit + no propagate inherit -> no inheritance
select acl_merge('{a/op/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

-- object inherit + no propagate inherit + inherit only -> no inheritance
select acl_merge('{a/opi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

--container inherit -> container inherit
select acl_merge('{a/c/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
                acl_merge                
-----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/hc/1=d}
(1 row)

--container inherit + inherit only -> container inherit
select acl_merge('{a/ci/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
                acl_merge                
-----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/hc/1=d}
(1 row)

-- container inherit + no propagate inherit -> no flags
select acl_merge('{a/cp/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- container inherit + no propagate inherit + inherit only -> no flags
select acl_merge('{a/cpi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- container inherit + object inherit -> container inherit + object inherit
select acl_merge('{a/co/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
                acl_merge                 
------------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/hco/1=d}
(1 row)

-- container inherit + object inherit + inherit only -> container inherit + object inherit
select acl_merge('{a/coi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
                acl_merge                 
------------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/hco/1=d}
(1 row)

-- container inherit + object inherit + no propagate inherit -> no flags
select acl_merge('{a/cop/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- container inherit + object inherit + no propagate inherit + inherit only -> no flags
select acl_merge('{a/copi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], true, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- object
-- no flags -> not inherited
select acl_merge('{a//1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

-- inherit only -> not inherited
select acl_merge('{a/i/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

-- object inherit -> no flags
select acl_merge('{a/o/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- object inherit + inherit only -> no flags
select acl_merge('{a/oi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- object inherit + no propagate inherit -> no flags
select acl_merge('{a/op/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- object inherit + no propagate inherit + inherit only -> no flags
select acl_merge('{a/opi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

--container inherit -> not inherited
select acl_merge('{a/c/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

--container inherit + inherit only -> not inherited
select acl_merge('{a/ci/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

-- container inherit + no propagate inherit -> not inherited
select acl_merge('{a/cp/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

-- container inherit + no propagate inherit + inherit only -> not inherited
select acl_merge('{a/cpi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

-- container inherit + object inherit -> no flags
select acl_merge('{a/co/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- container inherit + object inherit + inherit only -> no flags
select acl_merge('{a/coi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- container inherit + object inherit + no propagate inherit -> no flags
select acl_merge('{a/cop/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- container inherit + object inherit + no propagate inherit + inherit only -> no flags
select acl_merge('{a/copi/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
               acl_merge                
----------------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4,a/h/1=d}
(1 row)

-- skip inherited
select acl_merge('{a/h/1=d}'::ace_int8[], '{a//0=0,d//0=1,a//0=23,d//0=4}'::ace_int8[], false, false);
           acl_merge            
--------------------------------
 {a//0=0,d//0=1,a//0=23,d//0=4}
(1 row)

